File: src/encode/display_model.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""Recover a width-parametric display tree from three LuaTeX compilations.

The samples are ordinary serializer output at increasing ``\textwidth`` values.
No amsmath environment or node shape is special-cased: topology must match, and
every geometric scalar in the finished display tree must be affine across the
three widths.  The newest tree is retained and sparse ``*_rate`` derivatives
are attached to the fields which change.
"""

from __future__ import annotations

from typing import Any


NODE_GEOMETRY = (
    'width', 'height', 'depth', 'stretch', 'shrink', 'kern', 'shift',
    'glue_set', 'surround', 'm_a', 'm_b', 'm_c', 'm_d',
)
ITEM_GEOMETRY = ('display_width', 'display_indent', 'display_shift')
CHILD_LISTS = ('children', 'pre', 'post', 'replace')
RATE_KEYS = {f'{name}_rate' for name in NODE_GEOMETRY}
# The node types whose named field *is* horizontal space, and the field for each.
# Whether such a gap is internal to the formula or outer (centring) space is a
# rendering question, decided in the viewer against the reader's configured
# minimum; the encoder only records which fields are gaps at all.
GAP_FIELD_BY_TYPE = {'glue': 'width', 'kern': 'kern', 'math': 'surround'}


def _display_items(data: dict) -> list[dict]:
    items = [item for item in data.get('content', [])
             if item.get('kind') == 'display']
    for stream in data.get('streams', []):
        items.extend(item for item in stream.get('content', [])
                     if item.get('kind') == 'display')
    return items


def has_displays(data: dict) -> bool:
    return bool(_display_items(data))


def _node_topology_error(a: dict, b: dict, path: str) -> str | None:
    # How a box's glue was set is geometry, not identity: a formula TeX had to
    # shrink to fit beside its number at one measure is set at its natural
    # width at a wider one, the same tree in both. The affine check on
    # glue_set still rejects a setting that does not vary affinely.
    ignored = set(NODE_GEOMETRY) | RATE_KEYS | set(CHILD_LISTS) | {'leader', 'glue_sign', 'glue_order'}
    sa = {k: v for k, v in a.items() if k not in ignored}
    sb = {k: v for k, v in b.items() if k not in ignored}
    if sa != sb:
        return f'{path}: node identity changed ({sa!r} != {sb!r})'
    for key in CHILD_LISTS:
        aa, bb = a.get(key, []), b.get(key, [])
        if len(aa) != len(bb):
            return f'{path}.{key}: node count changed ({len(aa)} != {len(bb)})'
        for i, (na, nb) in enumerate(zip(aa, bb)):
            err = _node_topology_error(na, nb, f'{path}.{key}[{i}]')
            if err:
                return err
    la, lb = a.get('leader'), b.get('leader')
    if bool(la) != bool(lb):
        return f'{path}.leader: presence changed'
    if la:
        return _node_topology_error(la, lb, f'{path}.leader')
    return None


def _affine_error(a: float, b: float, c: float,
                  xa: int, xb: int, xc: int, *, integral: bool) -> float:
    predicted = b + (b - a) * (xc - xb) / (xb - xa)
    return abs(c - predicted)


def _node_affine_error(a: dict, b: dict, c: dict,
                       xa: int, xb: int, xc: int, path: str) -> str | None:
    for field in NODE_GEOMETRY:
        va, vb, vc = a.get(field, 0), b.get(field, 0), c.get(field, 0)
        err = _affine_error(va, vb, vc, xa, xb, xc,
                            integral=field not in {'glue_set', 'm_a', 'm_b', 'm_c', 'm_d'})
        tol = 3.0 if field not in {'glue_set', 'm_a', 'm_b', 'm_c', 'm_d'} \
            else 1e-6 * max(1.0, abs(va), abs(vb), abs(vc))
        if err > tol:
            return f'{path}.{field}: non-affine residual {err:g} (tolerance {tol:g})'
    for key in CHILD_LISTS:
        for i, (na, nb, nc) in enumerate(zip(a.get(key, []), b.get(key, []), c.get(key, []))):
            err = _node_affine_error(na, nb, nc, xa, xb, xc,
                                     f'{path}.{key}[{i}]')
            if err:
                return err
    if a.get('leader'):
        return _node_affine_error(a['leader'], b['leader'], c['leader'],
                                  xa, xb, xc, f'{path}.leader')
    return None


def check_samples(a: dict, b: dict, c: dict) -> tuple[bool, str | None]:
    """Return whether three increasing-width samples have one affine topology."""
    xa, xb, xc = (int(d.get('source_width', 0)) for d in (a, b, c))
    if not (0 < xa < xb < xc):
        return False, f'source widths are not strictly increasing: {xa}, {xb}, {xc}'
    da, db, dc = map(_display_items, (a, b, c))
    if not (len(da) == len(db) == len(dc)):
        return False, f'display count changed: {len(da)}, {len(db)}, {len(dc)}'
    for i, (ia, ib, ic) in enumerate(zip(da, db, dc)):
        err = _node_topology_error(ia['box'], ib['box'], f'display[{i}].box')
        if not err:
            err = _node_topology_error(ib['box'], ic['box'], f'display[{i}].box')
        if err:
            return False, err
        for field in ITEM_GEOMETRY:
            va, vb, vc = ia.get(field, 0), ib.get(field, 0), ic.get(field, 0)
            residual = _affine_error(va, vb, vc, xa, xb, xc, integral=True)
            if residual > 3:
                return False, f'display[{i}].{field}: non-affine residual {residual:g}'
        err = _node_affine_error(ia['box'], ib['box'], ic['box'],
                                 xa, xb, xc, f'display[{i}].box')
        if err:
            return False, err
    return True, None


def _attach_node_rates(recv: dict, previous: dict, newest: dict, dx: int) -> None:
    """Measure each field's slope between the two wider samples, and write it
    onto ``recv`` — the narrowest sample, which is the tree that is kept."""
    for field in NODE_GEOMETRY:
        delta = newest.get(field, 0) - previous.get(field, 0)
        if delta:
            recv[f'{field}_rate'] = delta / dx
    for key in CHILD_LISTS:
        for r, old_child, new_child in zip(recv.get(key, []),
                                           previous.get(key, []),
                                           newest.get(key, [])):
            _attach_node_rates(r, old_child, new_child, dx)
    if previous.get('leader'):
        _attach_node_rates(recv['leader'], previous['leader'], newest['leader'], dx)


def _mark_node_floors(oldest: dict, previous: dict, newest: dict, dx: int) -> None:
    # Only a gap can be floored, and only these node types carry one: a glue's
    # set width, a kern, and the space a math node surrounds itself with are the
    # horizontal space *between* pieces of a display. A box's ``width`` also
    # varies with the measure, but it is not space — it is however wide the
    # box's contents came out — so flooring it would fence off a measure the
    # contents never asked for. Other sampled fields (box height, vertical
    # shift, glue_set, transform coefficients) may legitimately cross zero and
    # define no horizontal validity boundary at all.
    field = GAP_FIELD_BY_TYPE.get(newest.get('type'))
    if field:
        va, vb, vc = oldest.get(field, 0), previous.get(field, 0), newest.get(field, 0)
        rate = (vc - vb) / dx
        if rate > 0 and va >= 0 and vb >= 0 and vc >= 0:
            oldest[f'{field}_floor'] = True
    for key in CHILD_LISTS:
        for na, nb, nc in zip(oldest.get(key, []), previous.get(key, []), newest.get(key, [])):
            _mark_node_floors(na, nb, nc, dx)
    if oldest.get('leader'):
        _mark_node_floors(oldest['leader'], previous['leader'], newest['leader'], dx)


def attach_model(oldest: dict, previous: dict, newest: dict) -> dict:
    """Attach sparse derivatives after ``check_samples`` succeeds, and return the
    **narrowest** sample as the document of record.

    The wider samples exist only to measure slopes with. Everything else in a
    compilation — paragraphs, colours, and any width a package baked into the
    page rather than into a display — belongs to the measure the document was
    actually written for. Keeping a wider sample would silently publish those at
    that width: a listings background, drawn as a rule across ``\linewidth``,
    would stretch as far as the widest probe reached. Slopes are derivatives, so
    they are the same law wherever it is anchored; anchoring it at the narrowest
    sample costs nothing and keeps the rest of the document honest.
    """
    xp, xn = int(previous['source_width']), int(newest['source_width'])
    dx = xn - xp
    for first_item, old_item, new_item in zip(_display_items(oldest),
                                               _display_items(previous),
                                               _display_items(newest)):
        for field in ITEM_GEOMETRY:
            delta = new_item.get(field, 0) - old_item.get(field, 0)
            if delta:
                first_item[f'{field}_rate'] = delta / dx
        _attach_node_rates(first_item['box'], old_item['box'], new_item['box'], dx)
        _mark_node_floors(first_item['box'], old_item['box'], new_item['box'], dx)
        # A natural-width display carries its centring in display_shift. Its
        # zero crossing is the point where the formula itself fills the column.
        shifts = (first_item.get('display_shift', 0), old_item.get('display_shift', 0),
                  new_item.get('display_shift', 0))
        shift_rate = (shifts[2] - shifts[1]) / dx
        if shift_rate > 0 and all(v >= 0 for v in shifts):
            first_item['display_shift_floor'] = True
    oldest['display_model'] = True
    return oldest


def anchor_model(first: dict, previous: dict, newest: dict) -> tuple[dict, int]:
    """``attach_model`` with the document of record fixed to ``first``: the sample
    at the width the document was written for, when *that* sample failed the
    affine check and the stable triple was found further up.

    What breaks the affine law between the document's own width and the wider
    samples is the width itself: a display skip that turns short, a kern that
    jumps, a box that shrinks only at the narrower measure. Anchoring the model
    at the first sample that happened to pass would publish the document as
    compiled at that wider width — different display skips, different
    breaks — and the page would not match the document at its own width, which
    is the one width it is expected to match exactly. So the slopes measured
    between the two wider samples are attached to ``first`` instead, to every
    display whose tree has the same shape there as in ``previous``; a display
    whose shape differs keeps its geometry fixed — it renders as TeX set it and
    does not reflow. Returns the document and the number of displays left fixed.
    """
    xp, xn = int(previous['source_width']), int(newest['source_width'])
    dx = xn - xp
    fixed = 0
    for i, (first_item, old_item, new_item) in enumerate(zip(_display_items(first),
                                                             _display_items(previous),
                                                             _display_items(newest))):
        if _node_topology_error(first_item['box'], old_item['box'], f'display[{i}].box'):
            fixed += 1
            continue
        for field in ITEM_GEOMETRY:
            delta = new_item.get(field, 0) - old_item.get(field, 0)
            if delta:
                first_item[f'{field}_rate'] = delta / dx
        _attach_node_rates(first_item['box'], old_item['box'], new_item['box'], dx)
        _mark_node_floors(first_item['box'], old_item['box'], new_item['box'], dx)
        shifts = (first_item.get('display_shift', 0), old_item.get('display_shift', 0),
                  new_item.get('display_shift', 0))
        shift_rate = (shifts[2] - shifts[1]) / dx
        if shift_rate > 0 and all(v >= 0 for v in shifts):
            first_item['display_shift_floor'] = True
    first['display_model'] = True
    return first, fixed

File: src/pipeline/display-model.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// A width-parametric display tree, recovered from three LuaTeX compilations of
// the document at increasing \textwidth.
//
// No amsmath environment or node shape is special-cased: the display trees of
// the samples must have one topology, and every geometric scalar in them must
// be affine across the three widths. The document of record keeps its own
// tree and gains sparse `*_rate` derivatives on the fields that change, and
// `*_floor` marks on the gaps that may close only so far. The viewer evaluates
// them at the reader's width. Every tree walk here is src/shared/tree.ts's.
import { walk1, some, walkParallel, type ChildKey } from '../shared/tree.ts';
import type { ContentItem, SerializerOutput, TexNode, FontInfo } from './nodes.ts';

const NODE_GEOMETRY = ['width', 'height', 'depth', 'stretch', 'shrink', 'kern', 'shift',
  'glue_set', 'surround', 'm_a', 'm_b', 'm_c', 'm_d'] as const;
const RATIO_FIELDS = new Set(['glue_set', 'm_a', 'm_b', 'm_c', 'm_d']);
const ITEM_GEOMETRY = ['display_width', 'display_indent', 'display_shift'] as const;
/** Not part of a node's identity: its geometry, the model's annotations, its
 *  descendants, and how its glue was set (a formula TeX had to shrink at one
 *  measure is set at its natural width at a wider one, the same tree). */
const NOT_IDENTITY = new Set<string>([
  ...NODE_GEOMETRY, ...NODE_GEOMETRY.map(f => `${f}_rate`), ...NODE_GEOMETRY.map(f => `${f}_floor`),
  'children', 'pre', 'post', 'replace', 'nobreak', 'leader', 'glue_sign', 'glue_order',
]);
/** The node types whose named field *is* horizontal space, and the field. */
const GAP_FIELD: Record<string, string> = { glue: 'width', kern: 'kern', math: 'surround' };

type Rec = Record<string, unknown>;
const num = (o: Rec, k: string): number => (o[k] as number | undefined) ?? 0;

// ── Python's formatting, for messages that read the same ────────────────────

/** Python's `format(x, 'g')`. */
export function pyG(x: number): string {
  if (!Number.isFinite(x)) return Number.isNaN(x) ? 'nan' : x > 0 ? 'inf' : '-inf';
  if (x === 0) return Object.is(x, -0) ? '-0' : '0';
  const exp = Math.floor(Math.log10(Math.abs(Number(x.toPrecision(6)))));
  if (exp < -4 || exp >= 6) {
    const [m, e] = x.toExponential(5).split('e');
    const mant = m.includes('.') ? m.replace(/0+$/, '').replace(/\.$/, '') : m;
    const en = Number(e);
    return `${mant}e${en < 0 ? '-' : '+'}${String(Math.abs(en)).padStart(2, '0')}`;
  }
  const s = x.toFixed(Math.max(0, 5 - exp));
  return s.includes('.') ? s.replace(/0+$/, '').replace(/\.$/, '') : s;
}

/** Python's repr() of JSON-shaped data. (A float that happens to be integral
 *  prints as an int: JSON cannot tell 1.0 from 1.) */
export function pyRepr(v: unknown): string {
  if (v === null || v === undefined) return 'None';
  if (v === true) return 'True';
  if (v === false) return 'False';
  if (typeof v === 'number') return Number.isInteger(v) ? String(v) : pyFloat(v);
  if (typeof v === 'string') {
    const q = v.includes("'") && !v.includes('"') ? '"' : "'";
    const body = v.replace(/\\/g, '\\\\').replace(/\n/g, '\\n').replace(/\r/g, '\\r').replace(/\t/g, '\\t')
      .replace(new RegExp(q, 'g'), `\\${q}`);
    return q + body + q;
  }
  if (Array.isArray(v)) return `[${v.map(pyRepr).join(', ')}]`;
  if (v instanceof Map) return `{${[...v].map(([k, x]) => `${pyRepr(k)}: ${pyRepr(x)}`).join(', ')}}`;
  return `{${Object.entries(v as Rec).map(([k, x]) => `${pyRepr(k)}: ${pyRepr(x)}`).join(', ')}}`;
}
function pyFloat(x: number): string {
  const s = String(x);
  const e = Math.floor(Math.log10(Math.abs(x)));
  if (e < -4 || e >= 16) {
    const [m, ex] = x.toExponential().split('e');
    const en = Number(ex);
    return `${m}e${en < 0 ? '-' : '+'}${String(Math.abs(en)).padStart(2, '0')}`;
  }
  return s.includes('e') ? x.toFixed(20).replace(/0+$/, '') : s;
}

// ── Items ───────────────────────────────────────────────────────────────────

/** The display items: the main flow's, then each stream's (not their wide
 *  forms – those are the model's output). */
export function displayItems(data: SerializerOutput): ContentItem[] {
  const out = data.content.filter(it => it.kind === 'display');
  for (const s of data.streams) out.push(...(s.content ?? []).filter(it => it.kind === 'display'));
  return out;
}

export const hasDisplays = (data: SerializerOutput): boolean => displayItems(data).length > 0;

// ── Shape and law ───────────────────────────────────────────────────────────

/** Why two trees are not the same tree (identity of every node, the number of
 *  nodes in every list, the presence of every leader), or null. */
export function topologyError(a: TexNode, b: TexNode, path: string): string | null {
  const identity = (n: TexNode) => Object.fromEntries(Object.entries(n).filter(([k]) => !NOT_IDENTITY.has(k)));
  return walkParallel<TexNode>([a, b], {
    node([x, y], p) {
      const sx = identity(x), sy = identity(y);
      if (!sameRecord(sx, sy)) return `${p}: node identity changed (${pyRepr(sx)} != ${pyRepr(sy)})`;
    },
    list(key: ChildKey, [la, lb], p) {
      if (key === 'leader') { if (la.length !== lb.length) return `${p}: presence changed`; }
      else if (la.length !== lb.length) return `${p}: node count changed (${la.length} != ${lb.length})`;
    },
  }, path);
}
function sameRecord(a: Rec, b: Rec): boolean {
  const ka = Object.keys(a), kb = Object.keys(b);
  return ka.length === kb.length && ka.every(k => k in b && JSON.stringify(a[k]) === JSON.stringify(b[k]));
}

const affineResidual = (a: number, b: number, c: number, xa: number, xb: number, xc: number): number =>
  Math.abs(c - (b + (b - a) * (xc - xb) / (xb - xa)));

/** Why three same-shaped trees are not affine in the width – the first field
 *  whose value at xc is not where the line through xa and xb puts it – or null. */
export function affineError(a: TexNode, b: TexNode, c: TexNode, xa: number, xb: number, xc: number, path: string,
                            tolSp = 3, tolRatio = 1e-6): string | null {
  return walkParallel<TexNode>([a, b, c], {
    node([na, nb, nc], p) {
      for (const f of NODE_GEOMETRY) {
        const va = num(na, f), vb = num(nb, f), vc = num(nc, f);
        const err = affineResidual(va, vb, vc, xa, xb, xc);
        const tol = RATIO_FIELDS.has(f) ? tolRatio * Math.max(1.0, Math.abs(va), Math.abs(vb), Math.abs(vc)) : tolSp;
        if (err > tol) return `${p}.${f}: non-affine residual ${pyG(err)} (tolerance ${pyG(tol)})`;
      }
    },
  }, path);
}

const sourceWidth = (d: SerializerOutput): number => Math.trunc(Number(d.source_width ?? 0));

/** Whether three increasing-width samples have one affine topology:
 *  [true, null] or [false, why not]. */
export function checkSamples(a: SerializerOutput, b: SerializerOutput, c: SerializerOutput): [boolean, string | null] {
  const [xa, xb, xc] = [a, b, c].map(sourceWidth);
  if (!(0 < xa && xa < xb && xb < xc)) return [false, `source widths are not strictly increasing: ${xa}, ${xb}, ${xc}`];
  const [da, db, dc] = [a, b, c].map(displayItems);
  if (!(da.length === db.length && db.length === dc.length))
    return [false, `display count changed: ${da.length}, ${db.length}, ${dc.length}`];
  for (let i = 0; i < da.length; i++) {
    const [ia, ib, ic] = [da[i], db[i], dc[i]];
    const err = topologyError(ia.box!, ib.box!, `display[${i}].box`) ?? topologyError(ib.box!, ic.box!, `display[${i}].box`);
    if (err) return [false, err];
    for (const f of ITEM_GEOMETRY) {
      const r = affineResidual(num(ia, f), num(ib, f), num(ic, f), xa, xb, xc);
      if (r > 3) return [false, `display[${i}].${f}: non-affine residual ${pyG(r)}`];
    }
    const e2 = affineError(ia.box!, ib.box!, ic.box!, xa, xb, xc, `display[${i}].box`);
    if (e2) return [false, e2];
  }
  return [true, null];
}

// ── Rates and floors ────────────────────────────────────────────────────────

/** Each field's slope between `previous` and `newest`, written onto `recv`'s
 *  corresponding node (`recv` may be `previous` itself). */
function attachRates(recv: TexNode, previous: TexNode, newest: TexNode, dx: number): void {
  walkParallel<TexNode>([recv, previous, newest], {
    node([r, p, n]) {
      for (const f of NODE_GEOMETRY) {
        const delta = num(n, f) - num(p, f);
        if (delta) r[`${f}_rate`] = delta / dx;
      }
    },
  });
}

/** Mark the gaps that were nonnegative throughout and open with the width:
 *  the viewer may floor them. Only a gap – a glue's set width, a kern, a math
 *  node's surround – is floored; a box's width varies too, but it is however
 *  wide its contents came out, not space. A box whose glue stretches more as
 *  the measure grows (a tabular*'s \extracolsep) is spaced by its glue_set,
 *  which closes at zero: its natural width. */
function markFloors(oldest: TexNode, previous: TexNode, newest: TexNode, dx: number): void {
  walkParallel<TexNode>([oldest, previous, newest], {
    node([o, p, n]) {
      const field = GAP_FIELD[n.type];
      if (field) {
        const va = num(o, field), vb = num(p, field), vc = num(n, field);
        if ((vc - vb) / dx > 0 && va >= 0 && vb >= 0 && vc >= 0) o[`${field}_floor`] = true;
      }
      if ((n.type === 'hlist' || n.type === 'vlist') && n.glue_sign === 1) {
        const [ga, gb, gc] = [o, p, n].map(d => (d.glue_set as number) || 0);
        if ((gc - gb) / dx > 0 && ga >= 0 && gb >= 0 && gc >= 0) o.glue_set_floor = true;
      }
    },
  });
}

function attachItem(first: ContentItem, old: ContentItem, next: ContentItem, dx: number): void {
  for (const f of ITEM_GEOMETRY) {
    const delta = num(next, f) - num(old, f);
    if (delta) first[`${f}_rate`] = delta / dx;
  }
  attachRates(first.box!, old.box!, next.box!, dx);
  markFloors(first.box!, old.box!, next.box!, dx);
  // A natural-width display carries its centring in display_shift; its zero
  // crossing is where the formula itself fills the column.
  const shifts = [first, old, next].map(d => num(d, 'display_shift'));
  if ((shifts[2] - shifts[1]) / dx > 0 && shifts.every(v => v >= 0)) first.display_shift_floor = true;
}

/** Attach the rates once checkSamples succeeds; the **narrowest** sample is
 *  the document of record (the wider ones exist to measure slopes with, and
 *  everything else a compilation holds belongs to the width the document was
 *  written for). */
export function attachModel(oldest: SerializerOutput, previous: SerializerOutput, newest: SerializerOutput): SerializerOutput {
  const dx = sourceWidth(newest) - sourceWidth(previous);
  const [f, o, n] = [oldest, previous, newest].map(displayItems);
  for (let i = 0; i < Math.min(f.length, o.length, n.length); i++) attachItem(f[i], o[i], n[i], dx);
  attachParagraphRates(oldest, previous, newest);
  oldest.display_model = true;
  return oldest;
}

/** attachModel with the document of record fixed to `first`, the sample at
 *  the document's own width, when that sample failed the affine check and the
 *  stable triple was found further up: the slopes of the stable pair go onto
 *  every display whose tree has the same shape at the document's width; one
 *  whose shape differs keeps its geometry (it renders as TeX set it). Returns
 *  the document and the number of displays left fixed. */
export function anchorModel(first: SerializerOutput, previous: SerializerOutput, newest: SerializerOutput): [SerializerOutput, number] {
  const dx = sourceWidth(newest) - sourceWidth(previous);
  const [f, o, n] = [first, previous, newest].map(displayItems);
  let fixed = 0;
  for (let i = 0; i < Math.min(f.length, o.length, n.length); i++) {
    if (topologyError(f[i].box!, o[i].box!, `display[${i}].box`)) { fixed++; continue; }
    attachItem(f[i], o[i], n[i], dx);
  }
  attachParagraphRates(first, previous, newest);
  first.display_model = true;
  return [first, fixed];
}

// ── A display set another way at the document's width ───────────────────────
// Such a display (amsmath moved its tag, or had no room to centre the body)
// gets two forms: its tree at the document's width, and the stable pair's own
// tree with its own rates (display_wide, anchored at display_wide_width), used
// from display_wide_from on – the width where TeX changes regime, found by
// compiling in between.

const WIDE_RESOLUTION_SP = 2 * 65536;           // bisect the switch width to within 2 pt

function displayAgrees(item: ContentItem, prev: ContentItem, next: ContentItem, x: number, xb: number, xc: number): boolean {
  if (topologyError(item.box!, prev.box!, 'box')) return false;
  for (const f of ITEM_GEOMETRY) if (affineResidual(num(item, f), num(prev, f), num(next, f), x, xb, xc) > 3) return false;
  return affineError(item.box!, prev.box!, next.box!, x, xb, xc, 'box') === null;
}

function clearRates(root: TexNode): void {
  walk1(root, n => {
    for (const k of Object.keys(n)) if (k.endsWith('_rate') || k.endsWith('_floor')) delete n[k];
  });
}

/** Font ids of `other`'s compilation → ids in `first`'s table, matched by
 *  file, name and size; a font only `other` loaded is added to `first`. */
function fontRemap(first: SerializerOutput, other: SerializerOutput): Map<string, string> {
  const key = (f: FontInfo) => JSON.stringify([f.filename, f.name, f.size_sp]);
  const byKey = new Map([...first.fonts].map(([id, f]) => [key(f), id] as const));
  const remap = new Map<string, string>();
  for (const [id, f] of other.fonts) {
    const k = key(f);
    if (!byKey.has(k)) {
      const fresh = String(Math.max(0, ...[...first.fonts.keys()].map(Number)) + 1);
      first.fonts.set(fresh, { ...f });
      byKey.set(k, fresh);
    }
    remap.set(id, byKey.get(k)!);
  }
  return remap;
}

/** Give each display of `first` (already anchored by anchorModel) that
 *  disagrees with the stable pair a wide form. `probe(width)` compiles the
 *  document at that \textwidth. Returns [displays given a wide form,
 *  compilations made]. */
export async function wideVariants(first: SerializerOutput, previous: SerializerOutput, newest: SerializerOutput,
                                   probe: (width: number) => Promise<SerializerOutput | null>): Promise<[number, number]> {
  const x0 = sourceWidth(first), xb = sourceWidth(previous), xc = sourceWidth(newest);
  const dx = xc - xb;
  const firsts = displayItems(first), prevs = displayItems(previous), news = displayItems(newest);
  if (!(firsts.length === prevs.length && prevs.length === news.length)) return [0, 0];
  const todo = firsts.map((_, i) => i).filter(i =>
    !displayAgrees(firsts[i], prevs[i], news[i], x0, xb, xc) && !some([prevs[i].box!], n => n.type === 'picture'));
  if (!todo.length) return [0, 0];
  const probes = new Map<number, SerializerOutput | null>();
  const sample = async (w: number) => {
    if (!probes.has(w)) probes.set(w, await probe(w));
    return probes.get(w)!;
  };
  // The switch lies in (lo, hi]: at lo the display is set the narrow way, at
  // hi on the wide law. All displays are bisected together, sharing
  // compilations.
  const lo = new Map(todo.map(i => [i, x0])), hi = new Map(todo.map(i => [i, xb]));
  const loData = new Map<number, ContentItem>();
  for (;;) {
    const open = todo.filter(i => hi.get(i)! - lo.get(i)! > WIDE_RESOLUTION_SP);
    if (!open.length) break;
    const w = Math.floor((lo.get(open[0])! + hi.get(open[0])!) / 2);
    const d = await sample(w);
    const items = d ? displayItems(d) : [];
    for (const i of open) {
      if (!(lo.get(i)! < w && w < hi.get(i)!)) continue;
      if (items.length === firsts.length && displayAgrees(items[i], prevs[i], news[i], w, xb, xc)) hi.set(i, w);
      else {
        lo.set(i, w);
        if (items.length === firsts.length) loData.set(i, items[i]);
      }
    }
  }
  const remap = fontRemap(first, previous);
  for (const i of todo) {
    const item = firsts[i], prevItem = prevs[i], newItem = news[i];
    const wide = structuredClone(prevItem);
    clearRates(wide.box!);
    for (const f of ITEM_GEOMETRY) {
      const delta = num(newItem, f) - num(prevItem, f);
      if (delta) wide[`${f}_rate`] = delta / dx;
    }
    attachRates(wide.box!, prevItem.box!, newItem.box!, dx);
    markFloors(wide.box!, prevItem.box!, newItem.box!, dx);
    walk1(wide.box!, n => {
      if (n.type === 'glyph' && remap.has(String(n.font))) n.font = Number(remap.get(String(n.font)));
    });
    for (const k of ['display_wide', 'display_wide_from', 'display_wide_width']) delete wide[k];
    item.display_wide = wide;
    item.display_wide_width = xb;
    item.display_wide_from = hi.get(i);
    // The narrow form's own law, where a compilation in its regime exists:
    // measured on the narrow side, not borrowed from the wide one.
    const near = loData.get(i);
    if (near !== undefined && !topologyError(item.box!, near.box!, 'box')) {
      const dn = lo.get(i)! - x0;
      clearRates(item.box!);
      for (const f of ITEM_GEOMETRY) {
        delete item[`${f}_rate`];
        const delta = num(near, f) - num(item, f);
        if (delta) item[`${f}_rate`] = delta / dn;
      }
      attachRates(item.box!, item.box!, near.box!, dn);
    }
  }
  return [todo.length, probes.size];
}

// ── Paragraphs: boxes set to the measure ────────────────────────────────────
// A box inside a paragraph can be as wide as the text: a tabular* spread over
// \textwidth, a figure scaled to \linewidth, a rule across the column. Such a
// box is modelled like a display, so it follows the reader's measure. Unlike a
// display, a paragraph that does not fit the law is no reason to reject a
// sample (a \parbox of text re-breaks at every width and never will): it keeps
// the geometry TeX gave it at the document's width.

const WIDE_BOX_SHARE = 0.5;       // a box at least this share of the text width is worth sampling for
// How far off the law a paragraph box may be and still follow it: a picture
// scaled to the measure gets its height from the width in TeX's integer
// arithmetic, some hundreds of sp off a straight line; ratios (glue_set, a
// scaling matrix) graphicx writes with five decimals. Displays keep 3 sp.
const PARAGRAPH_TOL_SP = 4096;
const PARAGRAPH_TOL_RATIO = 1e-4;

/** Whether a paragraph holds a box wide enough to have been set to the
 *  measure – reason enough to compile the document at other widths. */
export function hasWidthBoxes(data: SerializerOutput): boolean {
  const sw = Math.trunc(Number(data.source_width || 0));
  if (sw <= 0) return false;
  return data.paragraphs.some(p => (p.nodes ?? []).some(n =>
    (n.type === 'hlist' || n.type === 'vlist' || n.type === 'picture') && (((n.width as number) || 0) >= WIDE_BOX_SHARE * sw)));
}

export const wantsModel = (data: SerializerOutput): boolean => hasDisplays(data) || hasWidthBoxes(data);

const countRates = (root: TexNode): number => {
  let k = 0;
  walk1(root, n => { k += Object.keys(n).filter(key => key.endsWith('_rate')).length; });
  return k;
};

/** Rates on the nodes of each paragraph of `first` whose tree has one shape at
 *  the three widths and is affine across them. Returns how many paragraphs
 *  changed with the width and were given rates. */
export function attachParagraphRates(first: SerializerOutput, previous: SerializerOutput, newest: SerializerOutput): number {
  const [x0, xb, xc] = [first, previous, newest].map(sourceWidth);
  const [pa, pb, pc] = [first, previous, newest].map(d => d.paragraphs);
  if (!(pa.length === pb.length && pb.length === pc.length) || !(0 < x0 && x0 < xb && xb < xc)) return 0;
  const dx = xc - xb;
  let modelled = 0;
  for (let i = 0; i < pa.length; i++) {
    const [ta, tb, tc] = [pa[i], pb[i], pc[i]].map(p => ({ type: 'hlist', children: p.nodes ?? [] }) as TexNode);
    if (topologyError(ta, tb, 'para') || topologyError(tb, tc, 'para')) continue;
    if (affineError(ta, tb, tc, x0, xb, xc, 'para', PARAGRAPH_TOL_SP, PARAGRAPH_TOL_RATIO)) continue;
    const before = countRates(ta);
    attachRates(ta, tb, tc, dx);
    markFloors(ta, tb, tc, dx);
    if (countRates(ta) > before) modelled++;
  }
  return modelled;
}

